### PR TITLE
Add ComputeClient methods to manage functions

### PR DESCRIPTION
**Pull request by rjmello**
Add `.register_function()` and `.delete_function()` methods to the `ComputeClient` class.

----
📚 Documentation preview 📚: https://globus-sdk-python--1078.org.readthedocs.build/en/1078/

### CODE REVIEW

File: src/globus_sdk/_generate_init.py
Comment (sirosen) [inline src/globus_sdk/_generate_init.py:140]: I _think_ we should name these `ComputeFunctionDocument` and `ComputeFunctionMetadata`, although I could be convinced otherwise.

Right now, we have a very flat namespace for the service-related SDK exports -- `globus_sdk.X` for all `X`.
In most cases they're unambiguously associated with a given service, e.g., `StorageGatewayPolicies` is associated with GCS and couldn't have meaning with respect to anything else.

In cases where user confusion seems possible, we opt for longer names, e.g., `SearchScrollQuery` rather than `ScrollQuery`.

There's no hard-and-fast rule here, so there's nothing definitively _wrong_ with `FunctionDocument` and `FunctionMetadata` as names. The underlying problem we're looking at is that services haven't been given their own namespaces to populate freely -- and it's something I've been mulling for a few years without having any clear plan of action yet.

When I look at these names not only in light of the Globus ecosystem, but in comparison with the broader Python ecosystem, I think `FunctionMetadata` looks too much like the type of object one would expect from `inspect` or `types` in the stdlib -- and the SDK does contain some special values in the Python context, like `MISSING`, which make that kind of confusion possible.

File: src/globus_sdk/services/compute/data.py
Comment (sirosen) [inline src/globus_sdk/services/compute/data.py:39]: Should the defaults for these be that they are present with `null` values, or that they be omitted?

In newer document types and APIs we're trying to use `MISSING` instead of `None`. Very frequently, `None` has been used, historically, as a default where `MISSING` would be better. It doesn't apply to all APIs, however. `MISSING` isn't _inherently better_. It's just different in a way which we should consider.

For an example of usage, look at the GCS EndpointDocument type:
https://github.com/globus/globus-sdk-python/blob/6613e15a8928db39149b630441b630a73a1bcc23/src/globus_sdk/services/gcs/data/endpoint.py#L92

When the SDK is encoding a document as JSON to send, it _strips out_ `MISSING` values. So the effect is the difference between the following two encodings:
```python
# as written today
encode(
  FunctionDocument('foo', 'bar')
)
# {"function_name": "foo", "function_code": bar", "description": null, "metadata": null, "group": null, "public": False}

# if it were written using MISSING for defaults
encode(
  FunctionDocument('foo', 'bar')
)
# {"function_name": "foo", "function_code": bar"}
```

Note that I also assumed in this example that we make `public: bool | MissingType = MISSING` the default.

The difference comes into play most prominently with APIs which use a null field to indicate that a value should change / be removed in an update call. But it also matters here: if `public = False` is the default, then that's a client-side default. If `public = MISSING` is the default, then that's a server-side default.

---

If you're curious about how the behavior is implemented, here's MISSING:
https://github.com/globus/globus-sdk-python/blob/6613e15a8928db39149b630441b630a73a1bcc23/src/globus_sdk/utils.py#L25

And here's where MISSING gets filtered from payloads:
https://github.com/globus/globus-sdk-python/blob/6613e15a8928db39149b630441b630a73a1bcc23/src/globus_sdk/transport/encoders.py#L96-L99

(It also gets filtered from headers and query params.)

---

The exact behavior you will want depends on the behavior of the service, so I'm not advocating for a change so much as I'm making sure we discuss which is appropriate for the Compute API.

File: src/globus_sdk/services/compute/data.py
Comment (sirosen) [inline src/globus_sdk/services/compute/data.py:17]: Can we define defaults for these fields from the current execution environment?

The former looks like it's formatted `sys.version_info` data. I just tried out this little ditty and it seems to do the right thing:
```python
import sys


def fmt_version(v) -> str:
    if v.releaselevel == "final":
        return f"{v.major}.{v.minor}.{v.micro}"
    else:
        designator = {
            "candidate": "rc",
            "beta": "b",
            "alpha": "a",
        }[v.releaselevel]
        return f"{v.major}.{v.minor}.{v.micro}{designator}{v.serial}"


print(fmt_version(sys.version_info))
```

We could do something similar to pull the `globus-compute-sdk` version using importlib metadata.

---

This is as much a question about understanding of the service as it is about what we could do in this class.
When are callers expected to construct this object? Is it typically referring to the current runtime environment or are there significant cases in which it is meant to refer to some _other_ runtime when data was constructed?

File: src/globus_sdk/_testing/data/compute/delete_function.py
Comment (sirosen) [inline src/globus_sdk/_testing/data/compute/delete_function.py:11]: I'm curious, what does the 302 in this context mean?

File: changelog.d/20241008_214456_30907815+rjmello_compute_client_function_methods.rst
Comment (sirosen) [inline changelog.d/20241008_214456_30907815+rjmello_compute_client_function_methods.rst:6]: I don't feel strongly about phrasing, but I think these should be combined:

```suggestion
- Add ``ComputeClient.register_function()`` and ``ComputeClient.delete_function()``. (:pr:`NUMBER`)
```

Also, I think another bullet point is needed for the addition of helper classes.

File: src/globus_sdk/_generate_init.py
Comment (rjmello) [reply src/globus_sdk/_generate_init.py:140]: No objections from me. This will be more relevant for the endpoint methods because we need to differentiate between GCS and Compute endpoints.

File: src/globus_sdk/services/compute/data.py
Comment (rjmello) [reply src/globus_sdk/services/compute/data.py:39]: The register function route currently handles null and missing values the same, so I'm fine with the proposed change.

File: src/globus_sdk/_testing/data/compute/delete_function.py
Comment (rjmello) [reply src/globus_sdk/_testing/data/compute/delete_function.py:11]: It's an HTTP status code integer meant to indicate a successful deletion. This is a relic that we will change in the v3 API route.

File: src/globus_sdk/services/compute/data.py
Comment (sirosen) [reply src/globus_sdk/services/compute/data.py:39]: If they're equivalent (which is common for many APIs), I have a slight preference for MISSING. That way, if the service _adds_ a default behavior -- e.g., "default description is `<name> created by <user> on <date>`" -- the SDK picks it up seamlessly.

File: src/globus_sdk/services/compute/data.py
Comment (rjmello) [reply src/globus_sdk/services/compute/data.py:17]: It refers to the runtime environment when serializing the function. We currently set these values in the Compute SDK, where the serialization takes place. I prefer to leave it as-is for now.

Comment (sirosen) [reply src/globus_sdk/services/compute/data.py:23]: Okay. It's easier to add defaults than it is to take them away (the former is backwards compatible but the latter is not), so I see no reason to block on this. Mostly was just curious. :slightly_smiling_face:

**Review by sirosen — APPROVED** [review]
I assume we want to wait for a Compute maintainer to read and approve as well, so I'll refrain from merging for now.

**Review by khk-globus — APPROVED** [review]
I have not tried this out, but a static analysis looks good to me.

File: changelog.d/20241008_214456_30907815+rjmello_compute_client_function_methods.rst
Comment (sirosen) [reply changelog.d/20241008_214456_30907815+rjmello_compute_client_function_methods.rst:8]: I've wanted to enhance our doc build pipeline to give a rendered context for these. Because we're using `scriv` to avoid changelog conflicts, it's possible to... :bulb: :exclamation: 
We'll probably change style to use sphinx roles if I can make things work the way I think we can.

**Pull request merged**
